tools/src/bin/rust-indexer.rs

```diff
@@ -200,13 +200,142 @@ fn def_kind_to_human(kind: DefKind) -> &'static str {
     }
 }
 
-fn visit(
+/// Potentially non-helpful mapping of impl kind.
+fn impl_kind_to_human(kind: &ImplKind) -> &'static str {
+    match kind {
+        ImplKind::Inherent => "impl",
+        ImplKind::Direct => "impl for",
+        ImplKind::Indirect => "impl for ref",
+        ImplKind::Blanket => "impl for where",
+        _ => "impl for where deref"
+    }
+}
+
+/// Given two spans, create a new super-span that encloses them both if the files match.  If the
+/// files don't match, just return the first span as-is.
+fn union_spans(a: &data::SpanData, b: &data::SpanData) -> data::SpanData {
+    if a.file_name != b.file_name {
+        return a.clone();
+    }
+
+    let (start_byte, start_line, start_col) = if a.byte_start < b.byte_start {
+        (a.byte_start, a.line_start, a.column_start)
+    } else {
+        (b.byte_start, b.line_start, b.column_start)
+    };
+
+    let (end_byte, end_line, end_col) = if a.byte_end > b.byte_end {
+        (a.byte_end, a.line_end, a.column_end)
+    } else {
+        (b.byte_end, b.line_end, b.column_end)
+    };
+
+
+    data::SpanData {
+        file_name: a.file_name.clone(),
+        byte_start: start_byte, byte_end: end_byte,
+        line_start: start_line, line_end: end_line,
+        column_start: start_col, column_end: end_col
+    }
+}
+
+/// For the purposes of trying to figure out the actual effective nesting range of some type of
+/// definition, union its span (which just really covers the symbol name) plus the spans of all of
+/// its descendants.  This should end up with a sufficiently reasonable line value.  This is a hack.
+fn recursive_union_spans_of_def(def: &data::Def, file_analysis: &data::Analysis, defs: &Defs) -> data::SpanData {
+    let mut span = def.span.clone();
+    for id in &def.children {
+        // It should already be the case that the children are in the same krate, but better safe
+        // than sorry.
+        if id.krate != def.id.krate {
+            continue;
+        }
+        let kid = defs.get(file_analysis, *id);
+
+        if let Some(ref kid) = kid {
+            if kid.children.len() > 0 {
+                let rec_span = recursive_union_spans_of_def(kid, file_analysis, defs);
+                span = union_spans(&span, &rec_span);
+            } else {
+                span = union_spans(&span, &kid.span);
+            }
+        }
+    }
+
+    span
+}
+
+/// Given a list of ids of defs, run recursive_union_spans_of_def on all of them and union up the
+/// result.  Necessary for when dealing with impls.
+fn union_spans_of_defs(initial_span: &data::SpanData, ids: &Vec<data::Id>, file_analysis: &data::Analysis,
```

Comment:
nit: Please run rustfmt on this.

Also `&[data::Id]`, `&Vec` never makes sense.

Reply:
rustfmt wanted to make some unrelated changes so my plan is to rebase this patch stack as:
1. rustfmt run on tree without any changes.
2. squashed changes with rustfmt run again.

I'll also push all of my changes before undertaking that change so that you can see the deltas made without any rustfmt going on.

Reply:
Okay, I think I've done this now...